File: 25.k-个一组翻转链表.py
#
# @lc app=leetcode.cn id=25 lang=python3
#
# [25] K 个一组翻转链表
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
nxt = None
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if not head: return head
        a, b = head, head
        for _ in range(k):
            if not b:
                return head
            b = b.next
        new_node = self.reverseNode(a, b)
        a.next = self.reverseKGroup(b, k)
        return new_node

    def reverseNode(self, a, b):
        pre = None
        cur = a
        while cur != b:
            nxt = cur.next
            cur.next = pre
            pre = cur
            cur = nxt
        return pre


# @lc code=end

File: 25.k-个一组翻转链表.java
/*
 * @lc app=leetcode.cn id=25 lang=java
 *
 * [25] K 个一组翻转链表
 */

// @lc code=start

import javax.swing.plaf.TreeUI;

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode a,b;
        a = b = head;
        for(int i=0;i<k;i++){
            if(b==null)return head;
            b = b.next;
        }
        ListNode newNode = reverseLink(a, b);
        a.next = reverseKGroup(b, k);
        return newNode;
    }

    public ListNode reverseLink(ListNode head, ListNode tail){
        if(head.next==tail)return head;
        ListNode root = reverseLink(head.next, tail);
        head.next.next = head;
        head.next = null;
        return root;
    }
}
// @lc code=end

